src/csharp/Grpc.Tools/GeneratorServices.cs

```diff
@@ -55,31 +54,58 @@ protected bool GrpcOutputPossible(ITaskItem proto)
                 && !gsm.EqualNoCase("false");
         }
 
-        public abstract string[] GetPossibleOutputs(ITaskItem proto);
+        // Update OutputDir and GrpcOutputDir for the item and all subsequent
+        // targets using this item. This should only be done if the real
+        // output directories for protoc should be modified.
+        public virtual void PatchOutputDirectory(ITaskItem protoItem)
+        {
+            // Nothing to do
+        }
+
+        public abstract string[] GetPossibleOutputs(ITaskItem protoItem);
     };
 
     // C# generator services.
     internal class CSharpGeneratorServices : GeneratorServices
     {
         public CSharpGeneratorServices(TaskLoggingHelper log) : base(log) { }
 
+        public override void PatchOutputDirectory(ITaskItem protoItem)
+        {
+            string root = protoItem.GetMetadata(Metadata.ProtoRoot);
+            string proto = protoItem.ItemSpec;
+            string relative = DepFileUtil.GetRelativeDir(root, proto, Log);
+
+            string outdir = protoItem.GetMetadata(Metadata.OutputDir);
+            string pathStem = Path.Combine(outdir, relative);
+            protoItem.SetMetadata(Metadata.OutputDir, pathStem);
+
+            // Override outdir if GrpcOutputDir present, default to proto output.
+            string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
+            if (grpcdir != "")
+            {
+                pathStem = Path.Combine(grpcdir, relative);
+            }
+            protoItem.SetMetadata(Metadata.GrpcOutputDir, pathStem);
+        }
+
         public override string[] GetPossibleOutputs(ITaskItem protoItem)
         {
             bool doGrpc = GrpcOutputPossible(protoItem);
-            var outputs = new string[doGrpc ? 2 : 1];
-            string basename = Path.GetFileNameWithoutExtension(protoItem.ItemSpec);
+            string proto = protoItem.ItemSpec;
+            string basename = Path.GetFileNameWithoutExtension(proto);
 
+            var outputs = new string[doGrpc ? 2 : 1];
             string outdir = protoItem.GetMetadata(Metadata.OutputDir);
+
             string filename = LowerUnderscoreToUpperCamelProtocWay(basename);
             outputs[0] = Path.Combine(outdir, filename) + ".cs";
 
             if (doGrpc)
             {
-                // Override outdir if kGrpcOutputDir present, default to proto output.
                 string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
                 filename = LowerUnderscoreToUpperCamelGrpcWay(basename);
-                outputs[1] = Path.Combine(
-                    grpcdir != "" ? grpcdir : outdir, filename) + "Grpc.cs";
+                outputs[1] = Path.Combine(grpcdir, filename) + "Grpc.cs";
```

Comment:
Other than this line, are the changes in this method before this line needed? Also, this seems like a breaking change, how will users relying on existing behaviour be affected?

Reply:
Grpcdir will be set in PatchOutputDir. So the behavior stays the same in this case. Only real behavior change is that it uses relative paths now based on what is set for ProtoRoot. If ProtoRoot is not set specifically, it will be set in MSBuild to the relative path of the file and that will result in the old behavior.

Reply:
I made some changes to keep the diff minimal as requested.